src/docs/frontend/using-styled-components.mdx

```diff
@@ -0,0 +1,221 @@
+
+The best styles are the ones you don’t write - whenever possible use existing components.
+
+## Basics
+
+To use Style Components pull in `styled` and apply your css. For consistent spacing you should also use `space()`:
+
+```tsx
+import styled from '@emotion/styled';
+import {space} from 'sentry/styles/space';
+
+const Numeric = styled('span')`
+  font-variant-numeric: tabular-nums;
+`;
+
+const SideBySide = styled('div')`
+  display: flex;
+  gap: ${space(2)};
+  flex-wrap: wrap;
+  align-items: flex-start;
+`;
+
+export default ExampleComponent() {
+  return (
+    <SideBySide>
+      <p>a number: <Numeric>1</Numeric></p>
+      <p>and another: <Numeric>2</Numeric></p>
+    </SideBySide>
+  )
+}
+```
+
+## Theme
+
+Take constants (z-indexes, paddings, colors) from [props.theme](https://github.com/getsentry/sentry/blob/master/static/app/utils/theme.tsx)
+
+```tsx
+import styled from '@emotion/styled';
+import {space} from 'sentry/styles/space';
+
+const SomeComponent = styled('div')`
+  border-radius: 1.45em;
+  font-weight: bold;
+  z-index: ${p => p.theme.zIndex.modal};
+  padding: ${space(1)} ${space(2)};
+  border: 1px solid ${p => p.theme.borderLight};
+  color: ${p => p.theme.purple};
+  box-shadow: ${p => p.theme.dropShadowHeavy};
+`;
+```
+
+## Dynamic Properties
+
+<Alert level="info">
+  Most of the time dynamic properties are not needed!
+</Alert>
+
+You can pass in custom props to your custom component and use that to set styles. This does create overhead at runtime and should be avoided. What follows is an example of how to pass custom props, and how to avoid doing so.
+
+### Avoid Dynamic Props
+```tsx
+import styled from '@emotion/styled';
+
+// ❌ Try to avoid this
+const Label = styled('label')<{isDisabled: boolean; isError: boolean}>`
+  color: ${p => p.isDisabled
+    ? p.theme.gray200
+    : p.isError
+    ? p.theme.error400
+    : inherit
+  };
+`;
+
+return <Label disabled={false} isError={true}>There is an error</Label>;
+```
+
+Choose one of the strategies below instead.
+
+### Use Data Selectors
+
+Data selectors can be an easy way to set styles based on simple/pre-defined state. When defining these attributes don't forget that browser have built-in selectors like `:disabled`, `:focus`, `:first-child` and so on.
+
+```tsx
+import styled from '@emotion/styled';
+
+// ✅ Use a data attribute to pass in your state, and leverage CSS selectors:
+const Label = styled('label')`
+  color: inherit;
+  &[data-is-error="true"] {
+    color: ${p => p.theme.error400};
+  }
+  
+  &:disabled {
+    color: ${p => p.theme.gray200};
+  }
+`;
+
+return <Label disabled={false} data-is-error={true}>There is an error</Label>;
```

Comment:
Styling based on `data-` attributes seems spooky, no? It works, but it's a pretty surprising approach! There's not much precedent for it. Is this really better than regular props?

Reply:
I think the downside of regular props is that it has more of a runtime cost to re-serialize the styles and inject them into the page. So not only are we re-rendering the react component, but also messing with the styles, creating a lot of redundant-ish classes which have mostly the same properties inside. 

Using data selectors avoids the runtime css changes.

Also I've (rarely) seem examples where the props are passed through to be rendered, like this: 
```
const Thing = styled(props => <span {...props} />)<{hello: string}>``;
return <Thing hello="world" />;
```
So this would just make that valid html

Reply:
That makes sense, I'm just surprised because I haven't seen any precedent of people using `data-` attributes for styling, that's all!